pkg/ingester/stream.go

```diff
@@ -163,7 +163,11 @@ func (s *stream) Push(
 		return 0, []entryWithError{{entry: &logproto.Entry{}, e: ErrEntriesExist}}
 	}
 
-	var bytesAdded int
+	toStore, invalid := s.validateEntries(entries, isReplay)
+	if s.cfg.RateLimitWholeStream && hasRateLimitErr(invalid) {
```

Comment:
I think we can do this more efficiently. Currently, we're
1) Iterating through the entries for validation, calculating total bytes
2) iterating through all entries again to attach errors to them if they trip the rate limit
3) iterating them a third time here to see if one of them is rate limited

I think we can do this more efficiently, but happy to let it be in a separate followup PR.